src/osd/OSD.cc

```diff
@@ -9661,224 +9662,241 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
       return;
     }
   }
-  OpQueueItem item = sdata->pqueue->dequeue();
-  if (osd->is_stopping()) {
-    sdata->shard_lock.Unlock();
-    return;    // OSD shutdown, discard.
-  }
-  const auto token = item.get_ordering_token();
-  auto r = sdata->pg_slots.emplace(token, nullptr);
-  if (r.second) {
-    r.first->second = make_unique<OSDShardPGSlot>();
-  }
-  OSDShardPGSlot *slot = r.first->second.get();
-  dout(20) << __func__ << " " << token
-	   << (r.second ? " (new)" : "")
-	   << " to_process " << slot->to_process
-	   << " waiting " << slot->waiting
-	   << " waiting_peering " << slot->waiting_peering
-	   << dendl;
-  slot->to_process.push_back(std::move(item));
-  dout(20) << __func__ << " " << slot->to_process.back()
-	   << " queued" << dendl;
 
- retry_pg:
-  PGRef pg = slot->pg;
-
-  // lock pg (if we have it)
-  if (pg) {
-    // note the requeue seq now...
-    uint64_t requeue_seq = slot->requeue_seq;
-    ++slot->num_running;
+  list<Context *> oncommits;
+  if (!sdata->context_queue.empty()) {
+    sdata->context_queue.swap(oncommits);
```

Comment:
Why not finish() the contexts here?

Reply:
at this point,  it already lock(sdata->shard_lock)
a)context callback may take long time
b)context will get pg lock and will cause dead lock.